scipy/stats/_hypotests.py

```diff
@@ -1393,3 +1396,674 @@ def cramervonmises_2samp(x, y, method='auto'):
+        batch_iv = batch
+    else:
+        batch_iv = int(batch)
+        if batch != batch_iv or batch_iv <= 0:
+            raise ValueError("`batch` must be a positive integer or None.")
+
+    alternatives = {'two-sided', 'greater', 'less'}
+    alternative = alternative.lower()
+    if alternative not in alternatives:
+        raise ValueError(f"`alternative` must be in {alternatives}")
+
+    random_state = check_random_state(random_state)
+
+    return (data_iv, statistic, permutation_type, vectorized, permutations_int,
+            batch_iv, alternative, axis_int, random_state)
+
+
+# This is mostly to protect me while developing. Eventually we want to batch
+# the computations, so this will be less of an issue, but if it sounds like a
+# good feature, let me know.
+def _memory_check(data, permutations):
+    needed = sum([sample.size for sample in data]) * permutations * 8
+    available = 1e9  # psutil.virtual_memory()[1] # can't use in SciPy
+    if needed > 0.8 * available:
+        raise MemoryError("Not enough memory available. "
+                          "Consider reducing the batch size.")
+
+
+def permutation_test(data, statistic, *, permutation_type='both',
+                     vectorized=False, permutations=np.inf, batch=None,
+                     alternative="two-sided", axis=0, random_state=None):
+    r"""
+    Performs a permutation test of a given statistic on provided data.
+
+    For independent sample statistics, the null hypothesis is that the data are
+    randomly sampled from the same distribution.
+    For paired sample statistics, two null hypothesis can be tested:
+    that the data are paired at random or that the data are assigned to samples
+    at random.
+
+    Parameters
+    ----------
+    data : iterable of array-like
+        Contains the samples, each of which is an array of observations.
+        Dimensions of sample arrays must be compatible for broadcasting except
+        along `axis`.
+    statistic : callable
+        Statistic for which the p-value of the hypothesis test is to be
+        calculated. `statistic` must be a callable that accepts samples
+        as separate arguments (e.g. ``statistic(*data)``) and returns the
+        resulting statistic.
+        If `vectorized` is set ``True``, `statistic` must also accept a keyword
+        argument `axis` and be vectorized to compute the statistic along the
+        provided `axis` of the sample arrays.
+    permutation_type : {'both', 'sample', 'order'}
+        The type of permutations to be performed, in accordance with the
+        null hypothesis. The first two permutation types are for paired sample
+        statistics, in which all samples contain the same number of
+        observations and observations with corresponding indices along `axis`
+        are considered to be paired; the third is for independent sample
+        statistics.
+
+        - ``'samples'`` : observations are assigned to different samples
```

Comment:
this seems to be misaligned to the description `permutation_type : {'both', 'sample', 'order'}`

Reply:
I will correct the names of the values. The order is intended, though. `"both"` has to be first in the set above, as it is the default, but it is more obvious what "both" means if it is described after the other two.